src/wasm/wat-parser.cpp

```diff
@@ -596,16 +709,55 @@ template<typename Ctx> struct InstrParserCtx : TypeParserCtx<Ctx> {
     }
   }
 
-  InstrT makeUnreachable() { return builder.makeUnreachable(); }
-  InstrT makeNop() { return builder.makeNop(); }
-
-  InstrT makeI32Const(uint32_t c) { return builder.makeConst(Literal(c)); }
-  InstrT makeI64Const(uint64_t c) { return builder.makeConst(Literal(c)); }
-  InstrT makeF32Const(float c) { return builder.makeConst(Literal(c)); }
-  InstrT makeF64Const(double c) { return builder.makeConst(Literal(c)); }
+  Result<> makeUnreachable(Index pos) {
+    return push(pos, builder.makeUnreachable());
+  }
+  Result<> makeNop(Index pos) { return push(pos, builder.makeNop()); }
+  Result<> makeBinary(Index pos, BinaryOp op) {
+    auto rhs = pop(pos);
+    CHECK_ERR(rhs);
+    auto lhs = pop(pos);
+    CHECK_ERR(lhs);
```

Comment:
I think we can autogenerate this using `wasm-delegations-fields`?

Reply:
For some instructions yes, but not for others. For example, `select` needs to do some extra validation for the type parameter, and I don't see a good way of integrating that with `wasm-delegations-fields`. Note that the validation can't happen entirely in the general parsing function because general `ResultsT` might not have the necessary information. Other expressions like blocks and calls with variable numbers of children won't work with `wasm-delegations-fields` because those children aren't attached to the expression yet.

Reply:
Ah, I see, thanks...